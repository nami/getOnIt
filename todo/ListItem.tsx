import React, {useState} from 'react';
import {StyleSheet, View} from 'react-native';
import {
  Checkbox,
  IconButton,
  List,
  TextInput,
  Text,
  useTheme,
} from 'react-native-paper';
import {fonts} from '../theme';

type ListItemProps = {
  id: string;
  testId: string;
  value: string;
  completed: boolean;
  updateListItem: (listItemId: string, newValue: string) => void;
  renderListItemInactive: (listItemId: string) => void;
  toggleComplete: (listItemId: string) => void;
};

const ListItem = ({
  id,
  value,
  completed,
  renderListItemInactive,
  toggleComplete,
  updateListItem,
  testId,
}: ListItemProps) => {
  const theme = useTheme();
  const [updateMode, setUpdateMode] = useState(false);
  const [newText, setNewText] = useState('');
  const [showWarning, setShowWarning] = useState(false);

  const checkbox = () => {
    return (
      <>
        <Checkbox
          testID={`checkbox-${testId}`}
          status={completed ? 'checked' : 'unchecked'}
          uncheckedColor={theme.colors.primary}
          onPress={() => toggleComplete(id)}
        />
      </>
    );
  };

  // either display text component or input component
  const updateText = () => {
    return updateMode ? (
      <TextInput
        testID={'edit-input'}
        placeholder={value}
        onChangeText={text => setNewText(text)}
        style={styles.textInput}
        underlineColor={theme.colors.primary}
      />
    ) : (
      <Text
        testID={`val-${value}`}
        style={[fonts.medium, {color: theme.colors.tertiary}]}
        variant="bodyLarge">
        {value}
      </Text>
    );
  };

  // edit and delete buttons
  // change icon on update mode to signal that you can submit via icon press
  const editDelete = () => {
    const buttonSize = 20;
    return (
      <View style={styles.editDeleteContainer}>
        {showWarning ? (
          <Text style={styles.warningText} variant="labelSmall">
            Please input text...
          </Text>
        ) : (
          <></>
        )}
        {updateMode ? (
          <IconButton
            testID={`edit-mode-icon-${value}`}
            icon="pencil-circle-outline"
            iconColor={theme.colors.primary}
            onPress={() => {
              if (!newText) {
                setShowWarning(true);
                return;
              }
              updateListItem(id, newText);
              setUpdateMode(false);
              setShowWarning(false);
            }}
            style={styles.iconButton}
            size={buttonSize}
          />
        ) : (
          <IconButton
            testID={`nonedit-mode-icon-${value}`}
            iconColor={theme.colors.primary}
            icon="pencil-circle"
            onPress={() => setUpdateMode(true)}
            style={styles.iconButton}
            size={buttonSize}
          />
        )}
        <IconButton
          testID={`delete-icon-${value}`}
          icon="delete"
          iconColor={theme.colors.primary}
          onPress={() => {
            renderListItemInactive(id);
          }}
          style={styles.iconButton}
          size={buttonSize}
        />
      </View>
    );
  };

  return (
    <View>
      <List.Item
        testID={testId}
        id={id}
        title={updateText}
        style={styles.listItem}
        left={checkbox}
        onPress={() => {
          toggleComplete(id);
        }}
      />
      {editDelete()}
    </View>
  );
};

const styles = StyleSheet.create({
  listItem: {
    width: '90%',
    backgroundColor: 'white',
    borderRadius: 10,
    marginVertical: '2%',
    marginLeft: '5%',
  },
  editDeleteContainer: {
    paddingVertical: '2%',
    paddingHorizontal: '5%',
    height: 25,
    flexDirection: 'row',
    width: '100%',
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  iconButton: {
    width: '8%',
  },
  textInput: {
    height: 25,
  },
  warningText: {
    color: 'red',
    height: 16,
  },
});

export default ListItem;
