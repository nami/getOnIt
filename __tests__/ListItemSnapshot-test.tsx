import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import ListItem from '../todo/ListItem';

it('renders a ListItem using Snapshots', () => {
  expect(
    renderer.create(
      <ListItem
        id={'1234'}
        value={'A to-do'}
        completed={false}
        updateListItem={jest.fn}
        toggleComplete={jest.fn}
        renderListItemInactive={jest.fn}
      />
    )
  ).toMatchSnapshot();
});
