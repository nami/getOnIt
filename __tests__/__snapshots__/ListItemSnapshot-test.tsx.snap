// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a ListItem using Snapshots 1`] = `
<View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    id="1234"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        false,
        [
          {
            "paddingRight": 24,
            "paddingVertical": 8,
          },
          {
            "backgroundColor": "white",
            "borderRadius": 10,
            "marginLeft": "5%",
            "marginVertical": "2%",
            "width": "90%",
          },
        ],
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "marginVertical": 6,
          "width": "100%",
        }
      }
    >
      <View
        accessibilityLiveRegion="polite"
        accessibilityRole="checkbox"
        accessibilityState={
          {
            "busy": undefined,
            "checked": false,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 18,
              "padding": 6,
            },
          ]
        }
        testID="checkbox-undefined"
        uncheckedColor="rgba(103, 80, 164, 1)"
      >
        <View
          style={
            {
              "opacity": 0,
            }
          }
        >
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontSize": 24,
                },
                [
                  {
                    "lineHeight": 24,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰄬
          </Text>
        </View>
      </View>
      <View
        style={
          [
            {
              "paddingLeft": 16,
            },
            {
              "flex": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "400",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "PingFangSC-Medium",
                  "fontWeight": "400",
                },
                {
                  "color": "rgba(125, 82, 96, 1)",
                },
              ],
            ]
          }
          testID="val-A to-do"
        >
          A to-do
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 25,
        "justifyContent": "flex-end",
        "paddingHorizontal": "5%",
        "paddingVertical": "2%",
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderRadius": 18,
          "height": 36,
          "margin": 6,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": "8%",
        }
      }
      testID="nonedit-mode-icon-A to-do-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "rgba(121, 116, 126, 1)",
            "borderRadius": 18,
            "borderWidth": 0,
            "elevation": 0,
            "flex": 1,
            "overflow": "hidden",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="nonedit-mode-icon-A to-do-container"
      >
        <View
          accessibilityComponentType="button"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityTraits="button"
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          centered={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              [
                {
                  "alignItems": "center",
                  "flexGrow": 1,
                  "justifyContent": "center",
                },
                {
                  "borderRadius": 18,
                },
              ],
            ]
          }
          testID="nonedit-mode-icon-A to-do"
        >
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontSize": 20,
                },
                [
                  {
                    "lineHeight": 20,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰛿
          </Text>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderRadius": 18,
          "height": 36,
          "margin": 6,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": "8%",
        }
      }
      testID="delete-icon-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "rgba(121, 116, 126, 1)",
            "borderRadius": 18,
            "borderWidth": 0,
            "elevation": 0,
            "flex": 1,
            "overflow": "hidden",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="delete-icon-container"
      >
        <View
          accessibilityComponentType="button"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityTraits="button"
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          centered={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              [
                {
                  "alignItems": "center",
                  "flexGrow": 1,
                  "justifyContent": "center",
                },
                {
                  "borderRadius": 18,
                },
              ],
            ]
          }
          testID="delete-icon"
        >
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontSize": 20,
                },
                [
                  {
                    "lineHeight": 20,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰆴
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
